crates/agglayer-storage/src/storage/backup/mod.rs

```diff
@@ -0,0 +1,327 @@
+        loop {
+            tokio::select! {
+                _ = cancellation_token.cancelled() => {
+                    info!("Backup engine cancelled");
+                    break;
+                }
+                Some(request) = self.backup_request.recv() =>{
+                    self.create_new_backup(&request)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Restore the state database from the latest backup.
+    pub fn restore(path: &Path, db_path: &Path) -> Result<(), BackupError> {
+        let env = rocksdb::Env::new()?;
+        let opts = rocksdb::backup::BackupEngineOptions::new(path)?;
+
+        let mut engine = RocksBackupEngine::open(&opts, &env)?;
+
+        std::fs::create_dir_all(db_path).unwrap();
+
+        Ok(engine.restore_from_latest_backup(db_path, db_path, &RestoreOptions::default())?)
+    }
+
+    /// Restore the state database from the defined backup version.
+    pub fn restore_at(path: &Path, db_path: &Path, version: u32) -> Result<(), BackupError> {
+        let env = rocksdb::Env::new()?;
+        let opts = rocksdb::backup::BackupEngineOptions::new(path)?;
+
+        let mut engine = RocksBackupEngine::open(&opts, &env)?;
+
+        std::fs::create_dir_all(db_path).unwrap();
+
+        Ok(engine.restore_from_backup(db_path, db_path, &RestoreOptions::default(), version)?)
+    }
+
+    pub fn list_backups(path: &Path) -> Result<BackupReport, BackupError> {
+        let env = rocksdb::Env::new()?;
+
+        let mut report = BackupReport::default();
+
+        let opts = BackupEngineOptions::new(path.join("state"))?;
+        let engine = RocksBackupEngine::open(&opts, &env)?;
+
+        report.state(engine.get_backup_info());
+
+        let opts = BackupEngineOptions::new(path.join("pending"))?;
+        let engine = RocksBackupEngine::open(&opts, &env)?;
+
+        report.pending(engine.get_backup_info());
+
+        let epoch_path = path.join("epochs");
+        let mut epochs = (read_dir(&epoch_path)?)
+            .flatten()
+            .filter_map(|d| {
+                d.file_name()
+                    .to_string_lossy()
+                    .parse::<u64>()
+                    .map(|r| (r, d.path()))
+                    .ok()
+            })
+            .collect::<Vec<_>>();
+
+        epochs.sort_by(|(p, _), (n, _)| p.cmp(n));
+
+        for (epoch_number, path) in epochs {
+            let opts = BackupEngineOptions::new(path)?;
+            let engine = RocksBackupEngine::open(&opts, &env)?;
+            report.epochs(epoch_number, engine.get_backup_info());
+        }
+
+        Ok(report)
+    }
+}
+
+impl Drop for BackupEngine {
+    fn drop(&mut self) {
+        self.env.set_background_threads(0);
+        self.env.set_low_priority_background_threads(0);
+        self.env.set_high_priority_background_threads(0);
+        self.env.set_bottom_priority_background_threads(0);
+
+        self.env.join_all_threads();
+    }
```

Comment:
It's not quite clear to me what's happening here, particularly why the `set_*_threads` calls are required. A comment would be appreciated.

Reply:
It prevents `rocksdb` to spawn any more threads and wait for completion of the current ones.